In [1]:
import h5py
import imageio # Why?
import robosuite

[robosuite WARNING] No private macro file found! (macros.py:57)
[robosuite WARNING] It is recommended to use a private macro file (macros.py:58)
[robosuite WARNING] To setup, run: python /home/vishal/Volume_E/Active/Gap_year/grand-project-2025/deepak/robosuite/robosuite/scripts/setup_macros.py (macros.py:59)
[robosuite WARNING] Could not import robosuite_models. Some robots may not be available. If you want to use these robots, please install robosuite_models from source (https://github.com/ARISE-Initiative/robosuite_models) or through pip install. (__init__.py:30)
[robosuite WARNING] Could not load the mink-based whole-body IK. Make sure you install related import properly, otherwise you will not be able to use the default IK controller setting for GR1 robot. (__init__.py:40)


In [2]:
import dexmimicgen

In [51]:
import numpy as np

In [68]:
'''
Attributes are a critical part of what makes HDF5 a “self-describing” format. They are small named pieces of data attached directly to Group and Dataset objects. This is the official way to store metadata in HDF5.

Each Group or Dataset has a small proxy object attached to it, at <obj>.attrs. Attributes have the following properties:

They may be created from any scalar or NumPy array

Each attribute should be small (generally < 64k)

There is no partial I/O (i.e. slicing); the entire attribute must be read.

'''


data['data'].attrs["env_args"]

'{\n    "env_name": "TwoArmBoxCleanup",\n    "env_version": "1.5.1",\n    "type": 1,\n    "env_kwargs": {\n        "robots": [\n            "PandaDexRH",\n            "PandaDexLH"\n        ],\n        "controller_configs": {\n            "type": "BASIC",\n            "body_parts": {\n                "right": {\n                    "type": "OSC_POSE",\n                    "input_max": 1,\n                    "input_min": -1,\n                    "output_max": [\n                        0.05,\n                        0.05,\n                        0.05,\n                        0.5,\n                        0.5,\n                        0.5\n                    ],\n                    "output_min": [\n                        -0.05,\n                        -0.05,\n                        -0.05,\n                        -0.5,\n                        -0.5,\n                        -0.5\n                    ],\n                    "kp": 150,\n                    "damping_ratio": 1,\n      

In [4]:
dataset_path = "/home/vishal/Volume_E/Active/Gap_year/grand-project-2025/deepak/dexmimicgen/datasets/two_arm_box_cleanup.hdf5"
data = h5py.File(dataset_path, 'r')
print(data.keys())

<KeysViewHDF5 ['data']>


In [9]:
print("Number of keys: ", len(data['data'].keys()))

print("Keys in 'data': ", data['data'].keys())

Number of keys:  1016
Keys in 'data':  <KeysViewHDF5 ['demo_0', 'demo_1', 'demo_10', 'demo_100', 'demo_1000', 'demo_1001', 'demo_1002', 'demo_1003', 'demo_1004', 'demo_1005', 'demo_1006', 'demo_1007', 'demo_1008', 'demo_1009', 'demo_101', 'demo_1010', 'demo_1011', 'demo_1012', 'demo_1013', 'demo_1014', 'demo_1015', 'demo_102', 'demo_103', 'demo_104', 'demo_105', 'demo_106', 'demo_107', 'demo_108', 'demo_109', 'demo_11', 'demo_110', 'demo_111', 'demo_112', 'demo_113', 'demo_114', 'demo_115', 'demo_116', 'demo_117', 'demo_118', 'demo_119', 'demo_12', 'demo_120', 'demo_121', 'demo_122', 'demo_123', 'demo_124', 'demo_125', 'demo_126', 'demo_127', 'demo_128', 'demo_129', 'demo_13', 'demo_130', 'demo_131', 'demo_132', 'demo_133', 'demo_134', 'demo_135', 'demo_136', 'demo_137', 'demo_138', 'demo_139', 'demo_14', 'demo_140', 'demo_141', 'demo_142', 'demo_143', 'demo_144', 'demo_145', 'demo_146', 'demo_147', 'demo_148', 'demo_149', 'demo_15', 'demo_150', 'demo_151', 'demo_152', 'demo_153', 'dem

In [11]:
data['data']['demo_10']

<HDF5 group "/data/demo_10" (5 members)>

In [12]:
data['data']['demo_10'].keys()

<KeysViewHDF5 ['action_dict', 'actions', 'datagen_info', 'obs', 'states']>

In [52]:
print(data['data']['demo_10']['datagen_info'].keys())
for key in data['data']['demo_10']['datagen_info'].keys():
    print(f"{key}: {data['data']['demo_10']['datagen_info'][key]}")

print("\nMore about sub-task termination signals =============================  \n")

print(data['data']['demo_10']['datagen_info']['subtask_term_signals'].keys())
for key in data['data']['demo_10']['datagen_info']['subtask_term_signals'].keys():
    print(f"{key}: {data['data']['demo_10']['datagen_info']['subtask_term_signals'][key]}")

print("Example reading of subtask termination signals: ", data['data']['demo_10']['datagen_info']['subtask_term_signals']['lid_off_ground'][30:40])
# Are sub task termination signals always 0?
print("Sum of subtask termination signal (lid_off_ground): ", {sum(data['data']['demo_10']['datagen_info']['subtask_term_signals']['lid_off_ground'])})
non_zero_sigs = np.array(data['data']['demo_10']['datagen_info']['subtask_term_signals']['lid_off_ground'])
print("Non zero signals at indices: ", non_zero_sigs)

<KeysViewHDF5 ['eef_pose', 'gripper_action', 'object_poses', 'subtask_term_signals', 'target_pose']>
eef_pose: <HDF5 dataset "eef_pose": shape (226, 2, 4, 4), type "<f8">
gripper_action: <HDF5 dataset "gripper_action": shape (226, 2, 6), type "<f8">
object_poses: <HDF5 group "/data/demo_10/datagen_info/object_poses" (2 members)>
subtask_term_signals: <HDF5 group "/data/demo_10/datagen_info/subtask_term_signals" (1 members)>
target_pose: <HDF5 dataset "target_pose": shape (226, 2, 4, 4), type "<f8">

More about sub-task termination signals =============================  

<KeysViewHDF5 ['lid_off_ground']>
lid_off_ground: <HDF5 dataset "lid_off_ground": shape (226,), type "<i8">
Example reading of subtask termination signals:  [0 0 0 0 0 0 0 0 0 0]
Sum of subtask termination signal (lid_off_ground):  {90}
Non zero signals at indices:  [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 

# Notes on Data Set Structure
1. We are only given augmented demonstrations
2. We have around 1000 demonstrations for two_arm_box_cleanup dataset. Size around 5 GB
3. Consists of two sub-tasks - one reaching the lid, and the other picking the lid off the ground and placing it in its target positon

In [37]:
print("Keys in 'action_dict': ", data['data']['demo_9']['action_dict'].keys())
print(f"{data['data']['demo_9']['action_dict']['left_rel_rot_axis_angle']}")
print()
print(" ==== All keys and their values in demo_9 ==== ")
print()
for key in data['data']['demo_9'].keys():
    print(f"{key}: {data['data']['demo_9'][key]}")
print()
print(" ==== All keys and their values in action_dict of demo_9 ==== ")
print()
for key in data['data']['demo_9']['action_dict'].keys():
    print(f"{key}: {data['data']['demo_9']['action_dict'][key]}")

Keys in 'action_dict':  <KeysViewHDF5 ['left_gripper', 'left_rel_pos', 'left_rel_rot_6d', 'left_rel_rot_axis_angle', 'right_gripper', 'right_rel_pos', 'right_rel_rot_6d', 'right_rel_rot_axis_angle']>
<HDF5 dataset "left_rel_rot_axis_angle": shape (244, 3), type "<f4">

 ==== All keys and their values in demo_9 ==== 

action_dict: <HDF5 group "/data/demo_9/action_dict" (8 members)>
actions: <HDF5 dataset "actions": shape (244, 24), type "<f8">
datagen_info: <HDF5 group "/data/demo_9/datagen_info" (5 members)>
obs: <HDF5 group "/data/demo_9/obs" (21 members)>
states: <HDF5 dataset "states": shape (244, 103), type "<f8">

 ==== All keys and their values in action_dict of demo_9 ==== 

left_gripper: <HDF5 dataset "left_gripper": shape (244, 6), type "<f4">
left_rel_pos: <HDF5 dataset "left_rel_pos": shape (244, 3), type "<f4">
left_rel_rot_6d: <HDF5 dataset "left_rel_rot_6d": shape (244, 6), type "<f4">
left_rel_rot_axis_angle: <HDF5 dataset "left_rel_rot_axis_angle": shape (244, 3), type 

# Notes on trajectory data

- The dataset for task `two_arm_box_cleanup` consists of 1016 demonstrations.

- Although trajectory lengths vary among the initial set of 10 demonstrations (between 209 and 244), most demonstrations post 30 seem to have a size of about 238 waypoints/state-action pairs. 

In [15]:
data['data']['demo_10']['actions']


<HDF5 dataset "actions": shape (226, 24), type "<f8">

In [28]:
data['data']['demo_10']['actions'][0]

array([ 7.53327273e-02, -2.81475012e-02,  7.30194213e-02, -4.37843846e-03,
       -3.12801227e-02,  4.12576832e-04,  0.00000000e+00,  0.00000000e+00,
        0.00000000e+00,  0.00000000e+00,  0.00000000e+00,  1.57079633e+00,
       -7.66736512e-02, -4.66412965e-03, -2.63613784e-02, -2.01307777e-02,
        2.47150920e-02,  3.56377698e-02,  0.00000000e+00,  0.00000000e+00,
        0.00000000e+00,  0.00000000e+00,  0.00000000e+00,  1.57079633e+00])

In [29]:
for key in data['data']['demo_10']['action_dict'].keys():
    print(f"{key}: {data['data']['demo_10']['action_dict'][key][0]}")

left_gripper: [0.        0.        0.        0.        0.        1.5707964]
left_rel_pos: [-0.07667365 -0.00466413 -0.02636138]
left_rel_rot_6d: [ 0.99905974 -0.03587291  0.02434704  0.03537547  0.9991625   0.02056342]
left_rel_rot_axis_angle: [-0.02013078  0.02471509  0.03563777]
right_gripper: [0.        0.        0.        0.        0.        1.5707964]
right_rel_pos: [ 0.07533272 -0.0281475   0.07301942]
right_rel_rot_6d: [ 9.9951071e-01 -3.4403487e-04 -3.1275824e-02  4.8098154e-04
  9.9999034e-01  4.3712584e-03]
right_rel_rot_axis_angle: [-0.00437844 -0.03128012  0.00041258]


# Notes on Action Dict and Actions

Action dict consists of additional information which actions does not. 

`Action Dict` consists of:
- left_gripper: `<HDF5 dataset "left_gripper": shape (226, 6), type "<f4">`
- left_rel_pos: `<HDF5 dataset "left_rel_pos": shape (226, 3), type "<f4">`
- `left_rel_rot_6d`: `<HDF5 dataset "left_rel_rot_6d": shape (226, 6), type "<f4">`
- left_rel_rot_axis_angle: `<HDF5 dataset "left_rel_rot_axis_angle": shape (226, 3), type "<f4">`
- right_gripper: `<HDF5 dataset "right_gripper": shape (226, 6), type "<f4">`
- right_rel_pos: `<HDF5 dataset "right_rel_pos": shape (226, 3), type "<f4">`
- `right_rel_rot_6d`: `<HDF5 dataset "right_rel_rot_6d": shape (226, 6), type "<f4">`
- right_rel_rot_axis_angle: `<HDF5 dataset "right_rel_rot_axis_angle": shape (226, 3), type "<f4">`

Info not present in Actions is highlighted in Action_dict (rel_rot_6d arrays)

Whereas, action dictionary consists of a subset of this information:
- left_gripper: `<HDF5 dataset "left_gripper": shape (226, 6), type "<f4">`
- left_rel_pos: `<HDF5 dataset "left_rel_pos": shape (226, 3), type "<f4">`
- left_rel_rot_axis_angle: `<HDF5 dataset "left_rel_rot_axis_angle": shape (226, 3), type "<f4">`
- right_gripper: `<HDF5 dataset "right_gripper": shape (226, 6), type "<f4">`
- right_rel_pos: `<HDF5 dataset "right_rel_pos": shape (226, 3), type "<f4">`
- right_rel_rot_axis_angle: `<HDF5 dataset "right_rel_rot_axis_angle": shape (226, 3), type "<f4">`

In [56]:
print(f"Observations shape: {data['data']['demo_10']['obs'].keys()}")

Observations shape: <KeysViewHDF5 ['agentview_image', 'robot0_eef_pos', 'robot0_eef_quat', 'robot0_eef_quat_site', 'robot0_eye_in_hand_image', 'robot0_gripper_qpos', 'robot0_gripper_qvel', 'robot0_joint_pos', 'robot0_joint_pos_cos', 'robot0_joint_pos_sin', 'robot0_joint_vel', 'robot1_eef_pos', 'robot1_eef_quat', 'robot1_eef_quat_site', 'robot1_eye_in_hand_image', 'robot1_gripper_qpos', 'robot1_gripper_qvel', 'robot1_joint_pos', 'robot1_joint_pos_cos', 'robot1_joint_pos_sin', 'robot1_joint_vel']>


In [57]:
'''
Understanding images in the observation data
'''
image_data_keys = ['agentview_image', 'robot0_eye_in_hand_image', 'robot1_eye_in_hand_image']
for key in image_data_keys:
    print(f"Shape of {key}: {data['data']['demo_10']['obs'][key].shape}")
# print(f"I")

Shape of agentview_image: (226, 84, 84, 3)
Shape of robot0_eye_in_hand_image: (226, 84, 84, 3)
Shape of robot1_eye_in_hand_image: (226, 84, 84, 3)


In [63]:
'''
1. What is 'robot0_eef_quat_site'? What does it encode differently when compared to 'robot0_eef_quat'?
2. What are 'robot0_gripper_qpos', 'robot0_gripper_qvel'? How are they different from 'robot0_eef_pos', 'robot0_eef_quat'?
3. What are 'robot0_joint_pos_cos', 'robot0_joint_pos_sin'? Why do we need them, when we already have 'robot0_joint_pos'?
'''
print("Robot quat site: ", data['data']['demo_10']['obs']['robot0_eef_quat_site'][0])
print("Robot quat: ", data['data']['demo_10']['obs']['robot0_eef_quat'][0])
print("\n\n")
print("Robot gripper qpos: ", data['data']['demo_10']['obs']['robot0_gripper_qpos'][50])
print("Robot gripper qvel: ", data['data']['demo_10']['obs']['robot0_gripper_qvel'][50])
print("Robot joint pos: ", data['data']['demo_10']['obs']['robot0_joint_pos'][50])
print("Robot joint vel: ", data['data']['demo_10']['obs']['robot0_joint_vel'][50])
print("Robot eef pos: ", data['data']['demo_10']['obs']['robot0_eef_pos'][50])
print("Robot eef quat: ", data['data']['demo_10']['obs']['robot0_eef_quat'][50])
print("\n\n")

Robot quat site:  [-0.02326871  0.7010004   0.7127386   0.00779116]
Robot quat:  [ 0.99966443  0.01094427  0.02196266 -0.00830017]



Robot gripper qpos:  [ 1.30066807e+00  7.75580481e-02  1.82543191e-02  4.10195388e-03
 -5.08135294e-04  7.35969256e-04 -4.93038054e-04  6.49767654e-04
 -3.86284477e-04  8.49120038e-04 -2.42472530e-04  1.10449142e-03]
Robot gripper qvel:  [-3.96775096e-03  3.03610544e-04  6.20536393e-04  7.77580005e-05
  1.77170569e-03  7.21522975e-04  1.80967598e-03  7.47264181e-04
  1.89508896e-03  6.77992170e-04  1.98500623e-03  5.48458338e-04]
Robot joint pos:  [ 0.33953891  0.91302686  0.03717349 -1.92145061 -0.31936718  2.99751729
  1.37168468]
Robot joint vel:  [-0.06782659  0.23847568  0.20685003  0.31740933  0.29767983 -0.12474099
 -0.34644535]
Robot eef pos:  [-0.01392331 -0.00947829  0.94100618]
Robot eef quat:  [ 0.99540886  0.04293592  0.08343563 -0.01887321]





# Notes on Observations

1. What is 'robot0_eef_quat_site'? What does it encode differently when compared to 'robot0_eef_quat'?
   1. Robot quat site represents a site labelled for a robot end-effector that usually typically called 'tool-center-point', which is some distance offset from the gripper itself. 
2. What are 'robot0_gripper_qpos', 'robot0_gripper_qvel'? How are they different from 'robot0_eef_pos', 'robot0_eef_quat'?
3. What are 'robot0_joint_pos_cos', 'robot0_joint_pos_sin'? Why do we need them, when we already have 'robot0_joint_pos'?

In [71]:
data['data']['demo_10']['datagen_info']['object_poses'].keys()

<KeysViewHDF5 ['box', 'lid']>

In [72]:
data['data']['demo_10']['datagen_info']['object_poses']['box'], data['data']['demo_10']['datagen_info']['object_poses']['lid']

(<HDF5 dataset "box": shape (226, 4, 4), type "<f8">,
 <HDF5 dataset "lid": shape (226, 4, 4), type "<f8">)

In [73]:
data['data']['demo_10']['datagen_info'].keys()

<KeysViewHDF5 ['eef_pose', 'gripper_action', 'object_poses', 'subtask_term_signals', 'target_pose']>

In [74]:
'''Printing how the data looks for each of the key in datagen_info
'''
for key in data['data']['demo_10']['datagen_info'].keys():
    print(f"Key: {key}")
    print(f"Value: {data['data']['demo_10']['datagen_info'][key]}")

Key: eef_pose
Value: <HDF5 dataset "eef_pose": shape (226, 2, 4, 4), type "<f8">
Key: gripper_action
Value: <HDF5 dataset "gripper_action": shape (226, 2, 6), type "<f8">
Key: object_poses
Value: <HDF5 group "/data/demo_10/datagen_info/object_poses" (2 members)>
Key: subtask_term_signals
Value: <HDF5 group "/data/demo_10/datagen_info/subtask_term_signals" (1 members)>
Key: target_pose
Value: <HDF5 dataset "target_pose": shape (226, 2, 4, 4), type "<f8">


In [77]:
print(f"EEF Pose 0: {data['data']['demo_10']['datagen_info']['eef_pose'][0]}")
print(f"Target pose 0: {data['data']['demo_10']['datagen_info']['target_pose'][0]}")

print('\n')

print(f"EEF Pose 1: {data['data']['demo_10']['datagen_info']['eef_pose'][1]}")
print(f"Target pose 1: {data['data']['demo_10']['datagen_info']['target_pose'][1]}")

EEF Pose 0: [[[-0.99879573 -0.04372889 -0.02224577 -0.13490092]
  [-0.0215166  -0.0170755   0.99962266 -0.23224472]
  [-0.04409225  0.9988975   0.01611404  1.08629194]
  [ 0.          0.          0.          1.        ]]

 [[ 0.99576062 -0.09159092  0.00847903 -0.11707356]
  [ 0.00788088 -0.00689029 -0.99994521  0.25813644]
  [ 0.09164432  0.99577288 -0.00613927  1.0979786 ]
  [ 0.          0.          0.          1.        ]]]
Target pose 0: [[[-0.9979799  -0.05934271 -0.02268416 -0.13113429]
  [-0.02183613 -0.01490012  0.99965052 -0.2336521 ]
  [-0.05965997  0.99812646  0.01357421  1.08994291]
  [ 0.          0.          0.          1.        ]]

 [[ 0.99650978 -0.07923135  0.02628012 -0.12090724]
  [ 0.0264914   0.00161647 -0.99964773  0.25790323]
  [ 0.07916096  0.99685494  0.00370978  1.09666053]
  [ 0.          0.          0.          1.        ]]]


EEF Pose 1: [[[-0.99820464 -0.05739496 -0.0171265  -0.13117135]
  [-0.01632371 -0.01442688  0.99976267 -0.23540594]
  [-0.05762842 

In [ ]:
'''
Are actions really just delta between target and current poses?

Nope, actions are actually the exact differences between current pose and the reached pose (delta pose). Whereas,
the target pose is the controller's target for the next step, which makes it execute the delta pose. 
'''
